pkg/kubectl/rolling_updater.go

```diff
@@ -93,27 +96,421 @@ const (
+	if source != sourceId || !ok {
+		return nil, false, fmt.Errorf("Missing/unexpected annotations for controller %s, expected %s : %s", controller.Name, sourceId, annotations)
+	}
+	return existing, true, nil
 }
 
-// AlwaysAccept is an UpdateAcceptor which always accepts the controller.
-type AlwaysAccept struct{}
+// cleanupWithClients performs cleanup tasks after the rolling update. Update
+// process related annotations are removed from oldRc and newRc. The
+// CleanupPolicy on config is executed.
+func (r *RollingUpdater) cleanupWithClients(oldRc, newRc *api.ReplicationController, config *RollingUpdaterConfig) error {
+	// Clean up annotations
+	var err error
+	newRc, err = r.c.ReplicationControllers(r.ns).Get(newRc.Name)
+	if err != nil {
+		return err
+	}
+	delete(newRc.Annotations, sourceIdAnnotation)
+	delete(newRc.Annotations, desiredReplicasAnnotation)
 
-// Accept implements UpdateAcceptor.
-func (a *AlwaysAccept) Accept(*api.ReplicationController) error { return nil }
+	newRc, err = r.c.ReplicationControllers(r.ns).Update(newRc)
+	if err != nil {
+		return err
+	}
+	scalerClient := NewScalerClient(r.c)
+	if err = wait.Poll(config.Interval, config.Timeout, scalerClient.ControllerHasDesiredReplicas(newRc)); err != nil {
+		return err
+	}
+	newRc, err = r.c.ReplicationControllers(r.ns).Get(newRc.Name)
+	if err != nil {
+		return err
+	}
 
-// DefaultUpdaterAcceptor always accepts controllers.
-var DefaultUpdateAcceptor UpdateAcceptor = &AlwaysAccept{}
+	switch config.CleanupPolicy {
+	case DeleteRollingUpdateCleanupPolicy:
+		// delete old rc
+		fmt.Fprintf(config.Out, "Update succeeded. Deleting %s\n", oldRc.Name)
+		return r.c.ReplicationControllers(r.ns).Delete(oldRc.Name)
+	case RenameRollingUpdateCleanupPolicy:
+		// delete old rc
+		fmt.Fprintf(config.Out, "Update succeeded. Deleting old controller: %s\n", oldRc.Name)
+		if err := r.c.ReplicationControllers(r.ns).Delete(oldRc.Name); err != nil {
+			return err
+		}
+		fmt.Fprintf(config.Out, "Renaming %s to %s\n", newRc.Name, oldRc.Name)
+		return Rename(r.c, newRc, oldRc.Name)
+	case PreserveRollingUpdateCleanupPolicy:
+		return nil
+	default:
+		return nil
+	}
+}
 
-func LoadExistingNextReplicationController(c *client.Client, namespace, newName string) (*api.ReplicationController, error) {
+// func extractMaxValue is a helper to extract config max values as either
+// absolute numbers or based on percentages of the original rc.
+func extractMaxValue(field util.IntOrString, name string, original int) (int, error) {
+	switch field.Kind {
+	case util.IntstrInt:
+		if field.IntVal < 0 {
+			return 0, fmt.Errorf("%s must be >= 0", name)
+		}
+		return field.IntVal, nil
+	case util.IntstrString:
+		s := strings.Replace(field.StrVal, "%", "", -1)
+		v, err := strconv.Atoi(s)
+		if err != nil {
+			return 0, fmt.Errorf("invalid %s value %q: %v", name, field.StrVal, err)
+		}
+		if v < 0 || v > 100 {
+			return 0, fmt.Errorf("%s must be >= 0 and <= 100", name)
```

Comment:
maxSurge can be > 100%?
For ex: while updating 10 old replicas to 20 new replicas.